README.md
# VERSION

version v0.2.0

# ABSTRACT

A command-line utility using the GitHub API to generate useful reports that
can be used to audit various aspects of GitHub configuration.

# INSTALLATION

This module is only available from this repository and can be installed via
one of the following methods:

### cpanm

```
cpanm git@github.com:jrmash/perl-App-GitHub-Audit.git
```

### make

```
perl Makefile.PL
make
make test
make install
```

# COMMANDS

## branch-protection-settings

Generate a report containing a list of the branch protections defined for each
repository in the specified organization or user namespaces.

```perl
gh-audit branch-protection-settings --help
usage:
    gh-audit branch-protection-settings [long options...]
    gh-audit branch-protection-settings --user jrmash
    gh-audit branch-protection-settings --help

short description:
    Generate a report containing a list of pull requests merged to each repository's default branch, along with the
    list of users who approved them.

options:
    --user           Include resources owned by this user in the report. [Multiple]
    --org            Include resources owned by this org in the report. [Multiple]
    --report-format  The format of the report to generate. [Default:"csv"; Possible values: csv, tsv, xlsx]
    --report-name    The base name of the report file to create. [Default:"branch-protection-settings"]
    --help           Prints this usage information. [Flag]

available subcommands:
    help  Prints this usage information
```

## pull-request-approvers

Generates a report containing pull requests merged to the default branch of
each repository, along with a list of approving users.

```perl
gh-audit pull-request-approvers --help
usage:
    gh-audit pull-request-approvers [long options...]
    gh-audit pull-request-approvers --user jrmash
    gh-audit pull-request-approvers --help

short description:
    Generate a report containing a list of pull requests merged to each repository's default branch, along with the
    list of users who approved them.

options:
    --user           Include resources owned by this user in the report. [Multiple]
    --org            Include resources owned by this org in the report. [Multiple]
    --from-date      Include pull requests merged on or after this date. [Default:"1970-01-01T00:00:00"]
    --until-date     Include pull requests merged on before this date. [Default:"2022-09-25T21:52:58"]
    --report-format  The format of the report to generate. [Default:"csv"; Possible values: csv, tsv, xlsx]
    --report-name    The base name of the report file to create. [Default:"pull-request-approvers"]
    --help           Prints this usage information. [Flag]
```

# AUTHOR

J.R. Mash

# CONTRIBUTOR

J.R. Mash <2574997+jrmash@users.noreply.github.com>

# COPYRIGHT AND LICENSE

This software is Copyright (c) 2022 by J.R. Mash.

This is free software, licensed under:

```
The Apache License, Version 2.0, January 2004
```


Changes
Revision History for {{$dist->name}}

{{$NEXT}}

v0.2.0    2022-10-15 08:46:58-07:00 America/Los_Angeles
  * Add `branch-protection-settings` command

v0.1.0    2022-10-01 22:23:37-07:00 America/Los_Angeles
  * Initial version, unleashed upon an unsuspecting world.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.025.
use strict;
use warnings;

use 5.036000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Command-line utility to generate reports from the GitHub API",
  "AUTHOR" => "J.R. Mash",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "App-GitHub-Audit",
  "EXE_FILES" => [
    "bin/gh-audit"
  ],
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.036000",
  "NAME" => "App::GitHub::Audit",
  "PREREQ_PM" => {
    "Config::Any" => 0,
    "Const::Fast" => 0,
    "Data::Table" => 0,
    "Data::Table::Excel" => 0,
    "DateTime" => 0,
    "DateTime::Format::ISO8601" => 0,
    "IO::Interactive" => 0,
    "Moose::Role" => 0,
    "MooseX::AlwaysCoerce" => 0,
    "MooseX::App" => 0,
    "MooseX::App::Command" => 0,
    "MooseX::NiftyDelegation" => 0,
    "MooseX::StrictConstructor" => 0,
    "MooseX::UndefTolerant" => 0,
    "Net::GitHub::V3" => "1.05",
    "Path::Tiny" => 0,
    "Term::ANSIColor" => 0,
    "Term::ReadKey" => 0,
    "Text::WagnerFischer" => 0,
    "Types::DateTime" => 0,
    "Types::Standard" => 0,
    "boolean" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::More" => "0.94"
  },
  "VERSION" => "v0.2.0",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Config::Any" => 0,
  "Const::Fast" => 0,
  "Data::Table" => 0,
  "Data::Table::Excel" => 0,
  "DateTime" => 0,
  "DateTime::Format::ISO8601" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Interactive" => 0,
  "Moose::Role" => 0,
  "MooseX::AlwaysCoerce" => 0,
  "MooseX::App" => 0,
  "MooseX::App::Command" => 0,
  "MooseX::NiftyDelegation" => 0,
  "MooseX::StrictConstructor" => 0,
  "MooseX::UndefTolerant" => 0,
  "Net::GitHub::V3" => "1.05",
  "Path::Tiny" => 0,
  "Term::ANSIColor" => 0,
  "Term::ReadKey" => 0,
  "Test::CheckDeps" => "0.010",
  "Test::More" => "0.94",
  "Text::WagnerFischer" => 0,
  "Types::DateTime" => 0,
  "Types::Standard" => 0,
  "boolean" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

if ( $^O eq 'Windows' ) {
	$WriteMakefileArgs{PREREQ_PM}{'Win32::Console::ANSI'} = $FallbackPrereqs{'Win32::Console::ANSI'} = '0';
}

WriteMakefile(%WriteMakefileArgs);


lib/App/GitHub/Audit.pm
package App::GitHub::Audit;

use v5.36;
use boolean;
use strict;
use warnings;

use Const::Fast;
use MooseX::App ('Color', 'Term', 'Typo', 'Version');
use MooseX::AlwaysCoerce;
use MooseX::StrictConstructor;
use MooseX::UndefTolerant;
use Net::GitHub::V3;

const my $MX_APP_SCREEN_WIDTH => 118;

## =============================================================================
## Application customizations and overrides required to suit the project vision.
## =============================================================================
$MooseX::App::Utils::SCREEN_WIDTH = $MX_APP_SCREEN_WIDTH;

app_command_name { ($_[0] =~ s/_/-/gmrs )[0] };
app_namespace( 'App::GitHub::Audit::Command' );
app_permute( true );
app_usage(join("\n",
    'gh-audit <command> [options]',
    'gh-audit <command> --help',
));

has '+help_flag' => (
    cmd_aliases => [],
);

use namespace::autoclean (-except => 'new_with_command');
__PACKAGE__->meta->make_immutable();
1;

__END__

=pod

=head1 ABSTRACT

=head1 DESCRIPTION

=cut


lib/App/GitHub/Audit/Command.pm
package App::GitHub::Audit::Command;

use v5.36;
use boolean;
use strict;
use warnings;

use MooseX::App::Command;
use MooseX::AlwaysCoerce;
use MooseX::StrictConstructor;
use MooseX::UndefTolerant;

## =============================================================================
has 'github_api' => (
## =============================================================================
    init_arg        => undef,
    is              => 'ro',
    isa             => 'Net::GitHub::V3',
    lazy            => true,
    
    builder         => '_build_github_api',
);
sub _build_github_api($self) {
    my $api = Net::GitHub::V3->new();
    
    if (my $token = $ENV{'GITHUB_TOKEN'}) {
        $api->access_token( $ENV{'GITHUB_TOKEN'} );
    }
    elsif ($ENV{'GITHUB_ACCT'} && $ENV{'GITHUB_ACCT_PASS'}) {
        $api->login( $ENV{'GITHUB_ACCT'} );
        $api->pass( $ENV{'GITHUB_ACCT_PASS'} );
    }
    else {
        warn( "WARNING: GitHub credentials not found, using anonymous access.\n" );
    }

    return( $api );
}

## =============================================================================
has '_github_pulls_api' => (
## =============================================================================
    init_arg        => undef,
    is              => 'bare',
    isa             => 'Net::GitHub::V3::PullRequests',
    lazy            => true,
    
    builder         => '_build__github_pulls_api',
    handles         => [
        'next_pull',
        'next_review',
    ],
);
sub _build__github_pulls_api($self) {
    return( $self->github_api()->pull_request() );
}

## =============================================================================
has '_github_repos_api' => (
## =============================================================================
    init_arg        => undef,
    is              => 'bare',
    isa             => 'Net::GitHub::V3::Repos',
    lazy            => true,
    
    builder         => '_build__github_repos_api',
    handles         => [
        'branch_protection',
        'next_org_repo',
        'next_user_repo',
    ],
);
sub _build__github_repos_api($self) {
    return( $self->github_api()->repos() );
}

## =============================================================================
## App command customizations and overrides required to suit the project vision.
## =============================================================================
has '+help_flag' => (
    cmd_aliases => [],
);

use namespace::autoclean;
__PACKAGE__->meta->make_immutable();
1;

__END__

=pod

=head1 ABSTRACT

=head1 DESCRIPTION

=cut


# use DateTime;
# use MooseX::App::Command;
# use MooseX::AlwaysCoerce;
# use MooseX::NiftyDelegation;
# use MooseX::StrictConstructor;
# use MooseX::UndefTolerant;
# 
# use Types::Standard ('Enum');
# 
# extends 'GitHub::Audit';
# with (
#     'App::GitHub::Audit::Role::DateInputs',
#     'App::GitHub::Audit::Role::OwnerInputs',
#     'App::GitHub::Audit::Role::ReportGeneration',
# );
# 
# ## =============================================================================
# has '+from' => (
# ## =============================================================================
#     documentation   => 'Include pull requests merged on or after this date.',
# );
# 
# ## =============================================================================
# has '+until' => (
# ## =============================================================================
#     documentation   => 'Include pull requests merged on before this date.',
# );
# 
# ## =============================================================================
# has '+report_name' => (
# ## =============================================================================
#     default     => 'pull-request-approvers',
# );
# 
# ## =============================================================================
# sub run ($self) {
# ## =============================================================================
#     $self->add_report_data_column( undef, 'Merged Date');
#     $self->add_report_data_column( undef, 'Owner' );
#     $self->add_report_data_column( undef, 'Repository' );
#     $self->add_report_data_column( undef, 'PR Creator' );
#     $self->add_report_data_column( undef, 'PR Link' );
#     $self->add_report_data_column( undef, 'PR Approvers' );
#     $self->add_report_param_row( ['From Date', $self->from_datetime_rfc822] );
#     $self->add_report_param_row( ['To Date', $self->until_datetime_rfc822] );
# 
#     for my $owner ($self->orgs()) {
#         $self->_process_org( $owner );
#     }
#     for my $owner ($self->users()) {
#         $self->_process_user( $owner );
#     }
#     $self->write_report();
#     
#     return;
# }
# 
# ## =============================================================================
# sub _process_org($self, $owner) {
# ## =============================================================================
#     while (my $repo = $self->next_org_repo($owner)) {
#         $self->_process_repository( $repo );
#     }
#     return;
# }
# 
# ## =============================================================================
# sub _process_user($self, $owner) {
# ## =============================================================================
#     while (my $repo = $self->next_user_repo($owner)) {
#         $self->_process_repository( $repo );
#     }
#     return;
# }
# 
# ## =============================================================================
# sub _process_repository($self, $repo) {
# ## =============================================================================
#     my $repoBranch = $repo->{'default_branch'};
#     my $repoName   = $repo->{'name'};
#     my $repoOwner  = $repo->{'owner'}->{'login'};
#     
#     printf( "Processing: %s/%s\n", $repoOwner, $repoName );
#     
#     my @prQuery = ($repoOwner, $repoName, {
#         base      => $repoBranch,
#         sort      => 'update', direction => 'desc',
#         state     => 'closed',
#     });
#     
#     while (my $pr = $self->next_pull(@prQuery)) {
#         ## Closed pull requests without this field in output
#         ## were not merged to master, so skip them.
#         next unless ($pr->{'merged_at'}); 
#         
#         ## Merged pull requests with a merged date/time that
#         ## falls outside the from/until times are skipped.
#         my $prCreator = $pr->{'user'}->{'login'};
#         my $prMergedAt = $self->parse_datetime( $pr->{'merged_at'} );
#         my $prUrl = $pr->{'html_url'};
#         
#         if ($self->compare_from_datetime($prMergedAt) < 0) {
#             # printf( "Skipping ... %s\n", $prUrl );
#             next;
#         }
#         if ($self->compare_until_datetime($prMergedAt) > 0) {
#             # printf( "Skipping ... %s\n", $prUrl );
#             next;
#         }
#         
#         ## 
#         my @prApprovers = ();
#         my @prReviewQuery = ( $repoOwner, $repoName, $pr->{'number'} );
#         my @reportRow = ($prMergedAt, $repoOwner, $repoName, $prCreator, $prUrl);
# 
#         while ( my $r = $self->next_review(@prReviewQuery) ) {
#             unless ($r->{'state'} eq 'APPROVED') {
#                 # printf( "Skipping ... %s\n", $r->{'html_url'} );
#                 next;
#             }
#             push( @prApprovers, $r->{'user'}->{'login'} );
#         }
#         
#         $self->add_report_data_row( [@reportRow, join(', ', @prApprovers)] );
#     }
#     return;
# }
# 
# ## =============================================================================
# ## Command customizations and overrides required to suit the project vision.
# ## =============================================================================
# command_long_description(undef);
# command_usage(join("\n",
#     'gh-audit pull-request-approvers [long options...]',
#     'gh-audit pull-request-approvers --user jrmash',
#     'gh-audit pull-request-approvers --help',
# ));
# 


lib/App/GitHub/Audit/Role/Inputs/DateTime.pm
package App::GitHub::Audit::Role::Inputs::DateTime;

use v5.36;
use boolean;
use strict;
use warnings;

use DateTime;
use Moose::Role;
use MooseX::AlwaysCoerce;
use MooseX::NiftyDelegation;
use MooseX::UndefTolerant;

use Types::DateTime ('DateTimeUTC', 'Format');

## =============================================================================
=option B<--from-date> [C<Default: 1970-01-01T00:00:00Z>]

An L<ISO 8601|https://en.wikipedia.org/wiki/ISO_8601> formatted date / time
string to use as the start point for the query.

=cut
## -----------------------------------------------------------------------------
has 'from' => (
## =============================================================================
    cmd_flag        => 'from-date',
    cmd_position    => 100,
    cmd_type        => 'option',
    
    is              => 'ro',  ## Required for NiftyDelegation
    isa             => DateTimeUTC->plus_coercions( Format['ISO8601'] ),
    lazy            => true,
    traits          => [ 'AppOption', Nifty ],
    
    default         => '1970-01-01T00:00:00',
    handles         => {
        compare_from_datetime   => sub { DateTime->compare( $_[1], $_ ) },
        from_datetime_rfc822    => sub { $_->strftime('%a, %d %b %Y %H:%M:%S %Z') },
    },
);

## =============================================================================
=option B<--until-date> [C<Default: L<DateTime-E<gt>now()|DateTime/now>>]

An L<ISO 8601|https://en.wikipedia.org/wiki/ISO_8601> formatted date / time
string to use as the end point for the query.

=cut
## -----------------------------------------------------------------------------
has 'until' => (
## =============================================================================
    cmd_flag        => 'until-date',
    cmd_position    => 100,
    cmd_type        => 'option',
    
    is              => 'ro',  ## Required for NiftyDelegation
    isa             => DateTimeUTC->plus_coercions( Format['ISO8601'] ),
    lazy            => true,
    traits          => [ 'AppOption', Nifty ],
    
    default         => DateTimeUTC->assert_coerce('now')->stringify(),
    handles         => {
        compare_until_datetime  => sub { DateTime->compare( $_[1], $_ ) },
        until_datetime_rfc822   => sub { $_->strftime('%a, %d %b %Y %H:%M:%S %Z') },
    },
);

## =============================================================================
has '_datetime_parser' => (
## =============================================================================
    is              => 'bare',
    isa             => 'DateTime::Format::ISO8601',
    lazy            => true,
    
    default         => sub { DateTime::Format::ISO8601->new() },
    handles         => {
        parse_datetime              => 'parse_datetime',
    },
);

use namespace::autoclean;
1;

__END__

=pod

=head1 ABSTRACT

=head1 DESCRIPTION

=cut


lib/App/GitHub/Audit/Role/Inputs/RepoOwner.pm
package App::GitHub::Audit::Role::Inputs::RepoOwner;

use v5.36;
use boolean;
use strict;
use warnings;

use Moose::Role;
use MooseX::AlwaysCoerce;
use MooseX::UndefTolerant;

use Types::Standard ('ArrayRef', 'StrMatch');

## -----------------------------------------------------------------------------
has 'orgs' => (
## -----------------------------------------------------------------------------
    cmd_flag        => 'org',
    cmd_type        => 'option',
    documentation   => 'Include resources owned by this org in the report.',
    
    is              => 'bare',
    isa             => ArrayRef [ StrMatch [qr/^[[:alnum:]-]+$/msx] ],
    lazy            => true,
    traits          => [ 'AppOption', 'Array' ],
    
    default         => sub { [] },
    handles         => {
        orgs                        => 'elements',
    },
);

## -----------------------------------------------------------------------------
has 'users' => (
## -----------------------------------------------------------------------------
    cmd_flag        => 'user',
    cmd_type        => 'option',
    documentation   => 'Include resources owned by this user in the report.',
    
    is              => 'bare',
    isa             => ArrayRef [ StrMatch [qr/^[[:alnum:]-]+$/msx] ],
    lazy            => true,
    traits          => [ 'AppOption', 'Array' ],
    
    default         => sub { [] },
    handles         => {
        users                       => 'elements',
    },
);

use namespace::autoclean;
1;

__END__

=pod

=head1 ABSTRACT

=head1 DESCRIPTION

=cut


lib/App/GitHub/Audit/Role/Outputs/ReportFile.pm
package App::GitHub::Audit::Role::Outputs::ReportFile;

use v5.36;
use boolean;
use strict;
use warnings;

use Const::Fast;
use Data::Table;
use Data::Table::Excel ('tables2xlsx');
use Moose::Role;
use MooseX::AlwaysCoerce;
use MooseX::NiftyDelegation;
use MooseX::UndefTolerant;
use Path::Tiny;

use Types::DateTime ('DateTimeUTC');
use Types::Standard ('Enum', 'InstanceOf', 'Str');

const my $EXCEL_TABLE_COLORS => [9, 44, 56];

## =============================================================================
has '_report_data' => (
## =============================================================================
    is              => 'ro',
    isa             => InstanceOf['Data::Table'],
    lazy            => true,
    
    builder         => '_build__report_data',
    handles         => {
        add_report_data_column      => 'addCol',
        add_report_data_row         => 'addRow',
        get_report_data_row_count   => 'nofRow',
    },
);
sub _build__report_data($self) {
    return( Data::Table->new([], [], Data::Table::ROW_BASED) );
}

## =============================================================================
has '_report_params' => (
## =============================================================================
    is              => 'ro',
    isa             => InstanceOf['Data::Table'],
    lazy            => true,
    
    builder         => '_build__report_params',
    handles         => {
        add_report_param_row        => 'addRow',
        get_report_param_row_count  => 'nofRow',
    },
);
sub _build__report_params($self) {
    my $table = Data::Table->new( [], [], Data::Table::ROW_BASED );
    
    $table->addCol( undef, 'Param' );
    $table->addCol( undef, 'Value' );
    $table->addRow( ['Created', DateTimeUTC->assert_coerce('now')] );
    
    return( $table );
}



## =============================================================================
has 'report_format' => (
## =============================================================================
    cmd_flag        => 'report-format',
    cmd_position    => 1000,
    cmd_type        => 'option',
    documentation   => 'The format of the report to generate.',
    
    is              => 'ro',
    isa             => Enum[ 'csv', 'tsv', 'xlsx' ],
    lazy            => true,
    traits          => [ 'AppOption' ],
    
    default         => 'csv',
);

## =============================================================================
has 'report_name' => (
## =============================================================================
    cmd_flag        => 'report-name',
    cmd_position    => 1000,
    cmd_type        => 'option',
    documentation   => 'The base name of the report file to create.',
    
    is              => 'ro',
    isa             => Str,
    lazy            => true,
    traits          => [ 'AppOption', 'String' ],
    
    default         => 'github-audit',
);

## =============================================================================
=method B<write_report>

=cut
## -----------------------------------------------------------------------------
sub write_report($self) {
## =============================================================================
    my $file = join( '.', $self->report_name(), $self->report_format() );
    my $filePath = path( $file );
    
    if ($self->report_format() eq 'csv') {
        $filePath->spew( $self->_report_params()->csv(), "\n" );
        $filePath->append( $self->_report_data()->csv() );
    }
    elsif ($self->report_format() eq 'tsv') {
        $filePath->spew( $self->_report_params()->tsv(), "\n" );
        $filePath->append( $self->_report_data()->tsv() );
    }
    elsif ($self->report_format() eq 'xlsx') {
        tables2xlsx( $file, [$self->_report_params(), $self->_report_data()], undef, [$EXCEL_TABLE_COLORS, undef] );
    }
    
    return;
}

use namespace::autoclean;
1;

__END__

=pod

=head1 ABSTRACT

=head1 DESCRIPTION

=cut


lib/App/GitHub/Audit/Command/branch_protection_settings.pm
package App::GitHub::Audit::Command::branch_protection_settings;

use v5.36;
use boolean;
use strict;
use warnings;

use MooseX::App::Command;
use MooseX::AlwaysCoerce;
use MooseX::StrictConstructor;
use MooseX::UndefTolerant;

extends 'App::GitHub::Audit::Command';
with (
    'App::GitHub::Audit::Role::Inputs::RepoOwner',
    'App::GitHub::Audit::Role::Outputs::ReportFile',
);

## =============================================================================
has '+report_name' => (
## =============================================================================
    default     => 'branch-protection-settings',
);

## =============================================================================
sub run ($self) {
## =============================================================================
    $self->add_report_data_column( undef, 'Owner' );
    $self->add_report_data_column( undef, 'Repository' );
    $self->add_report_data_column( undef, 'Branch' );
    $self->add_report_data_column( undef, 'Required Approvals' );

    for my $owner ($self->orgs()) {
        $self->_process_org( $owner );
    }
    for my $owner ($self->users()) {
        $self->_process_user( $owner );
    }
    $self->write_report();

    return;
}

## =============================================================================
sub _process_org($self, $owner) {
## =============================================================================
    while (my $repo = $self->next_org_repo($owner)) {
        $self->_process_repository( $repo );
    }
    return;
}

## =============================================================================
sub _process_user($self, $owner) {
## =============================================================================
    while (my $repo = $self->next_user_repo($owner)) {
        $self->_process_repository( $repo );
    }
    return;
}

## =============================================================================
sub _process_repository($self, $repo) {
## =============================================================================
    my $repoBranch = $repo->{'default_branch'};
    my $repoName   = $repo->{'name'};
    my $repoOwner  = $repo->{'owner'}->{'login'};

    printf( "Processing: %s/%s\n", $repoOwner, $repoName );

    ## Catch any errors that might come up when we try to get the branch
    ## protection settings. This may only be required due to an issue in
    ## the NET::GitHub module with the response parsing, but I will look
    ## at that later.
    eval {
        my $protection = $self->branch_protection($repoOwner, $repoName, $repoBranch);

        $self->add_report_data_row([
            $repoOwner,
            $repoName,
            $repoBranch,
            $protection->{'required_pull_request_reviews'}->{'required_approving_review_count'},
        ]);
    } or do {
        warn( "WARNING: Branch protections for '$repoName' have not been enablled.\n" );
    };
    return;
}

## =============================================================================
## App command customizations and overrides required to suit the project vision.
## =============================================================================
command_long_description(undef);
command_usage(join("\n",
    'gh-audit branch-protection-settings [long options...]',
    'gh-audit branch-protection-settings --user jrmash',
    'gh-audit branch-protection-settings --help',
));

use namespace::autoclean;
__PACKAGE__->meta->make_immutable();
1;

__END__

=pod

=head1 ABSTRACT

Generate a report containing a list of the branch protections defined for each
repository in the specified organization or user namespaces.

=head1 DESCRIPTION

    gh-audit branch-protection-settings --help
    usage:
        gh-audit branch-protection-settings [long options...]
        gh-audit branch-protection-settings --user jrmash
        gh-audit branch-protection-settings --help

    short description:
        Generate a report containing a list of pull requests merged to each repository's default branch, along with the
        list of users who approved them.

    options:
        --user           Include resources owned by this user in the report. [Multiple]
        --org            Include resources owned by this org in the report. [Multiple]
        --report-format  The format of the report to generate. [Default:"csv"; Possible values: csv, tsv, xlsx]
        --report-name    The base name of the report file to create. [Default:"branch-protection-settings"]
        --help           Prints this usage information. [Flag]

    available subcommands:
        help  Prints this usage information

=cut


lib/App/GitHub/Audit/Command/pull_request_approvers.pm
package App::GitHub::Audit::Command::pull_request_approvers;

use v5.36;
use boolean;
use strict;
use warnings;

use MooseX::App::Command;
use MooseX::AlwaysCoerce;
use MooseX::NiftyDelegation;
use MooseX::StrictConstructor;
use MooseX::UndefTolerant;

use Types::Standard ('Enum');

extends 'App::GitHub::Audit::Command';
with (
    'App::GitHub::Audit::Role::Inputs::DateTime',
    'App::GitHub::Audit::Role::Inputs::RepoOwner',
    'App::GitHub::Audit::Role::Outputs::ReportFile',
);

## =============================================================================
has '+from' => (
## =============================================================================
    documentation   => 'Include pull requests merged on or after this date.',
);

## =============================================================================
has '+until' => (
## =============================================================================
    documentation   => 'Include pull requests merged on before this date.',
);

## =============================================================================
has '+report_name' => (
## =============================================================================
    default     => 'pull-request-approvers',
);

## =============================================================================
sub run ($self) {
## =============================================================================
    $self->add_report_data_column( undef, 'Merged Date');
    $self->add_report_data_column( undef, 'Owner' );
    $self->add_report_data_column( undef, 'Repository' );
    $self->add_report_data_column( undef, 'PR Creator' );
    $self->add_report_data_column( undef, 'PR Link' );
    $self->add_report_data_column( undef, 'PR Approvers' );
    $self->add_report_param_row( ['From Date', $self->from_datetime_rfc822] );
    $self->add_report_param_row( ['To Date', $self->until_datetime_rfc822] );

    for my $owner ($self->orgs()) {
        $self->_process_org( $owner );
    }
    for my $owner ($self->users()) {
        $self->_process_user( $owner );
    }
    $self->write_report();

    return;
}

## =============================================================================
sub _process_org($self, $owner) {
## =============================================================================
    while (my $repo = $self->next_org_repo($owner)) {
        $self->_process_repository( $repo );
    }
    return;
}

## =============================================================================
sub _process_user($self, $owner) {
## =============================================================================
    while (my $repo = $self->next_user_repo($owner)) {
        $self->_process_repository( $repo );
    }
    return;
}

## =============================================================================
sub _process_repository($self, $repo) {
## =============================================================================
    my $repoBranch = $repo->{'default_branch'};
    my $repoName   = $repo->{'name'};
    my $repoOwner  = $repo->{'owner'}->{'login'};

    printf( "Processing: %s/%s\n", $repoOwner, $repoName );

    my @prQuery = ($repoOwner, $repoName, {
        base      => $repoBranch,
        sort      => 'update', direction => 'desc',
        state     => 'closed',
    });

    while (my $pr = $self->next_pull(@prQuery)) {
        ## Closed pull requests without this field in output
        ## were not merged to master, so skip them.
        next unless ($pr->{'merged_at'}); 

        ## Merged pull requests with a merged date/time that
        ## falls outside the from/until times are skipped.
        my $prCreator = $pr->{'user'}->{'login'};
        my $prMergedAt = $self->parse_datetime( $pr->{'merged_at'} );
        my $prUrl = $pr->{'html_url'};

        if ($self->compare_from_datetime($prMergedAt) < 0) {
            # printf( "Skipping ... %s\n", $prUrl );
            next;
        }
        if ($self->compare_until_datetime($prMergedAt) > 0) {
            # printf( "Skipping ... %s\n", $prUrl );
            next;
        }

        ## 
        my @prApprovers = ();
        my @prReviewQuery = ( $repoOwner, $repoName, $pr->{'number'} );
        my @reportRow = ($prMergedAt, $repoOwner, $repoName, $prCreator, $prUrl);

        while ( my $r = $self->next_review(@prReviewQuery) ) {
            unless ($r->{'state'} eq 'APPROVED') {
                # printf( "Skipping ... %s\n", $r->{'html_url'} );
                next;
            }
            push( @prApprovers, $r->{'user'}->{'login'} );
        }

        $self->add_report_data_row( [@reportRow, join(', ', @prApprovers)] );
    }
    return;
}

## =============================================================================
## App command customizations and overrides required to suit the project vision.
## =============================================================================
command_long_description(undef);
command_usage(join("\n",
    'gh-audit pull-request-approvers [long options...]',
    'gh-audit pull-request-approvers --user jrmash',
    'gh-audit pull-request-approvers --help',
));

use namespace::autoclean;
__PACKAGE__->meta->make_immutable();
1;

__END__

=pod

=head1 ABSTRACT

Generate a report containing a list of pull requests merged to each repository's
default branch, along with the list of users who approved them.

=head1 DESCRIPTION

    gh-audit pull-request-approvers --help
    usage:
        gh-audit pull-request-approvers [long options...]
        gh-audit pull-request-approvers --user jrmash
        gh-audit pull-request-approvers --help

    short description:
        Generate a report containing a list of pull requests merged to each repository's default branch, along with the
        list of users who approved them.

    options:
        --user           Include resources owned by this user in the report. [Multiple]
        --org            Include resources owned by this org in the report. [Multiple]
        --from-date      Include pull requests merged on or after this date. [Default:"1970-01-01T00:00:00"]
        --until-date     Include pull requests merged on before this date. [Default:"2022-09-25T21:52:58"]
        --report-format  The format of the report to generate. [Default:"csv"; Possible values: csv, tsv, xlsx]
        --report-name    The base name of the report file to create. [Default:"pull-request-approvers"]
        --help           Prints this usage information. [Flag]

    available subcommands:
        help  Prints this usage information

=cut
